### PR TITLE
Xingpeng/add grounding metric

### PR DESCRIPTION
**Pull request by msxingpeng**

### CODE REVIEW

File: visionmetrics/grounding/recall.py [hunk 1]
```diff
@@ -9,6 +9,8 @@ class Recall(Metric):
 
     def __init__(self, iou_thresh=0.5, k=1):
         super().__init__()
+        if iou_thresh < 0 or iou_thresh > 1:
+            raise ValueError(f"iou_thresh must be in [0, 1], got {iou_thresh}.")
```
Comment (pj-ms) [inline visionmetrics/grounding/recall.py:13]: nit: extra line after raise

File: visionmetrics/grounding/recall.py [hunk 2]
```diff
@@ -39,6 +41,16 @@ def update(self, predictions, targets):
                      [([target_phrase1, target_phrase2, ...], [[target_bbox, target_bbox, ...], [target_bbox, target_bbox, ...], ...]), ...], type: list of tuple.
                      "target_bboxes" is a list of list of bbox (top, left, bottom, right) in absolute scale.
         """
+        if (len(predictions)!=len(targets)):
+            raise ValueError(f"Number of predictions and targets must be equal, got {len(predictions)} and {len(targets)}.")
```
Comment (pj-ms) [inline visionmetrics/grounding/recall.py:45]: extra line after raise

File: visionmetrics/grounding/recall.py [hunk 3]
```diff
@@ -39,6 +41,16 @@ def update(self, predictions, targets):
                      [([target_phrase1, target_phrase2, ...], [[target_bbox, target_bbox, ...], [target_bbox, target_bbox, ...], ...]), ...], type: list of tuple.
                      "target_bboxes" is a list of list of bbox (top, left, bottom, right) in absolute scale.
         """
+        if (len(predictions)!=len(targets)):
```
Comment (pj-ms) [inline visionmetrics/grounding/recall.py]: remove ()

File: visionmetrics/grounding/recall.py [hunk 4]
```diff
@@ -39,6 +41,16 @@ def update(self, predictions, targets):
...
+            raise ValueError(f"Number of predictions and targets must be equal, got {len(predictions)} and {len(targets)}.")
+        for pred, target in zip(predictions, targets):
+            pred_phrases, pred_bboxes = pred
+            target_phrases, target_bboxes = target
+            if(len(pred_phrases)!=len(pred_bboxes)):
```
Comment (pj-ms) [inline visionmetrics/grounding/recall.py]: remove ()

File: visionmetrics/grounding/recall.py [hunk 5]
```diff
@@ -9,6 +9,8 @@ class Recall(Metric):
 
     def __init__(self, iou_thresh=0.5, k=1):
         super().__init__()
+        if iou_thresh < 0 or iou_thresh > 1:
+            raise ValueError(f"iou_thresh must be in [0, 1], got {iou_thresh}.")
         self.iou_thresh = iou_thresh
         self.topk = k
```
Comment (pj-ms) [inline visionmetrics/grounding/recall.py:15]: range check for k

File: visionmetrics/grounding/recall.py (hunk 1 above)
Comment (pj-ms) [reply visionmetrics/grounding/recall.py:13]: Your check is (0, 1), while the range is [0, 1] (inclusive) here

File: visionmetrics/grounding/recall.py [hunk 6]
```diff
@@ -39,6 +41,16 @@ def update(self, predictions, targets):
                      [([target_phrase1, target_phrase2, ...], [[target_bbox, target_bbox, ...], [target_bbox, target_bbox, ...], ...]), ...], type: list of tuple.
                      "target_bboxes" is a list of list of bbox (top, left, bottom, right) in absolute scale.
         """
+        if (len(predictions) != len(targets)):
+            raise ValueError(f"Number of predictions and targets must be equal, got {len(predictions)} and {len(targets)}.")
+        for pred, target in zip(predictions, targets):
```
Comment (pj-ms) [inline visionmetrics/grounding/recall.py:46]: If this grounding specific, shouldn't it be enough for the ground truth wrt text being the text span instead of a real string?

File: visionmetrics/grounding/recall.py [hunk 7]
```diff
@@ -9,6 +9,8 @@ class Recall(Metric):
 
     def __init__(self, iou_thresh=0.5, k=1):
         super().__init__()
+        if iou_thresh < 0 or iou_thresh > 1:
```
Comment (pj-ms) [inline visionmetrics/grounding/recall.py:12]: There is a IOU implementation https://torchmetrics.readthedocs.io/en/stable/detection/intersection_over_union.html

Not quite sure if it is a good choice to re-use, or a test can be implemented using it?

File: visionmetrics/grounding/recall.py (hunk 6 above)
Comment (msxingpeng) [reply visionmetrics/grounding/recall.py:51]: Ideally yes, the string itself does not matter if the text span matches.

File: visionmetrics/grounding/recall.py (hunk 7 above)
Comment (msxingpeng) [reply visionmetrics/grounding/recall.py]: Not the same, IOU in torchmetrics computes the the IOU between a list of predictions and a list of boxes and return an average IOU,

Here we are computing the IOU between N's bbox and M's bbox, and return NxM matrix. 

```python
import torch
from torchmetrics.detection import IntersectionOverUnion
preds = [{"boxes": torch.tensor([[300.00, 100.00, 315.00, 150.00],[300.00, 100.00, 315.00, 150.00], [100.00, 200.00, 315.00, 450.00]]), 
          "scores": torch.tensor([0.2, 0.3, 0.2]),
          "labels": torch.tensor([5, 6, 7])}]

target = [{"boxes": torch.tensor([[300.00, 100.00, 315.00, 150.00], [100.00, 200.00, 315.00, 450.00]]),"labels": torch.tensor([5, 7 ])}]
metric = IntersectionOverUnion()
metrics = metric(preds, target)
print(metrics)

# result: {'iou': tensor(0.3333)}

class test_iou():
    def __init__(self):
        pass

    def _box_area(self, boxes):
        assert boxes.ndim == 2 and boxes.shape[-1] == 4
        return (boxes[:, 2] - boxes[:, 0]) * (boxes[:, 3] - boxes[:, 1])

    def _box_iou(self, boxes1, boxes2):
        area1 = self._box_area(boxes1)
        area2 = self._box_area(boxes2)
        lt = torch.maximum(boxes1[:, None, :2], boxes2[:, :2])
        rb = torch.minimum(boxes1[:, None, 2:], boxes2[:, 2:])
        wh = (rb - lt).clip(min=0)
        inter = wh[:, :, 0] * wh[:, :, 1]
        union = area1[:, None] + area2 - inter
        iou = inter / union
        return iou

tester = test_iou()
iou = tester._box_iou(preds[0]['boxes'], target[0]['boxes'])
print(iou)

# result: tensor([[1., 0.], [1., 0.],[0., 1.]])
```

File: setup.cfg [hunk 8]
```diff
@@ -1,6 +1,6 @@
 [metadata]
 name = visionmetrics
-version = 0.0.10
+version = 0.0.11
```
Comment (pj-ms) [inline setup.cfg:3]: looks like 0.0.10 is not released, no need to bump the version

File: visionmetrics/grounding/recall.py [hunk 9]
```diff
@@ -39,6 +56,18 @@ def update(self, predictions, targets):
                      [([target_phrase1, target_phrase2, ...], [[target_bbox, target_bbox, ...], [target_bbox, target_bbox, ...], ...]), ...], type: list of tuple.
                      "target_bboxes" is a list of list of bbox (top, left, bottom, right) in absolute scale.
         """
+        if len(predictions) != len(targets):
+            raise ValueError(f"Number of predictions and targets must be equal, got {len(predictions)} and {len(targets)}.")
+
+        for pred, target in zip(predictions, targets):
+            pred_phrases, pred_bboxes = pred
+            target_phrases, target_bboxes = target
+            if len(pred_phrases) != len(pred_bboxes):
```
Comment (pj-ms) [inline visionmetrics/grounding/recall.py:65]: hard to locate which one triggered this error in the list

File: visionmetrics/grounding/recall.py [hunk 10]
```diff
@@ -9,6 +9,12 @@ class Recall(Metric):
 
     def __init__(self, iou_thresh=0.5, k=1):
         super().__init__()
+        if not (0 <= iou_thresh <= 1):
+            raise ValueError(f"iou_thresh must be in [0, 1], got {iou_thresh}.")
+
+        if not isinstance(k, int) or not (1 <= k <= 5):
+            raise ValueError(f"k must be an integer in [1, 5], got {k}.")
```
Comment (pj-ms) [inline visionmetrics/grounding/recall.py:16]: why is 5 the upper bound?

Comment (msxingpeng) [reply visionmetrics/grounding/recall.py]: Changed to positive integer

File: setup.cfg (hunk 8 above)
Comment (msxingpeng) [reply setup.cfg]: Reverted

File: visionmetrics/grounding/recall.py (hunk 3 above)
Comment (msxingpeng) [reply visionmetrics/grounding/recall.py]: removed

File: visionmetrics/grounding/recall.py (hunk 2 above)
Comment (msxingpeng) [reply visionmetrics/grounding/recall.py:60]: added

**Review by pj-ms — APPROVED** [review]
